### PR TITLE
MAINT: add qrvs method to NumericalInversePolynomial in scipy.stats

### PR DESCRIPTION
**Pull request by chrisb83**
#### Reference issue
N/A

#### What does this implement/fix?
Add qrvs method to `NumericalInversePolynomial` (PINV) that already exists for `NumericalInverseHermite` (HINV)

#### Additional information
The implementation is exactly the same as for HINV. I tried to avoid too much overlap of the code by creating a common function for input validation and by adding parametrize to the test cases

### CODE REVIEW

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 1]
```diff
@@ -373,6 +373,26 @@ cdef double[::1] _validate_pv(pv) except *:
...
+        d = qmc_engine.d if d is None else d
+    elif qmc_engine is None:
+        d = 1 if d is None else d
+        qmc_engine = stats.qmc.Halton(d)
+    else:
+        message = ("`qmc_engine` must be an instance of "
+                    "`scipy.stats.qmc.QMCEngine` or `None`.")
+        raise ValueError(message)
+
+    return qmc_engine, d
```
Comment (mdhaber) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:393]: I can confirm that this is simply moved from below.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 2]
```diff
@@ -2082,22 +2196,7 @@ cdef class NumericalInverseHermite(Method):
...
-            d = qmc_engine.d if d is None else d
-        elif qmc_engine is None:
-            d = 1 if d is None else d
-            qmc_engine = stats.qmc.Halton(d)
-        else:
-            message = ("`qmc_engine` must be an instance of "
-                       "`scipy.stats.qmc.QMCEngine` or `None`.")
-            raise ValueError(message)
-
+        qmc_engine, d = _validate_qmc_input(qmc_engine, d)
```
Comment (mdhaber) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:2199]: This looks equivalent to before.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 3]
```diff
@@ -1614,6 +1634,100 @@ cdef class NumericalInversePolynomial(Method):
         return UError(max_error, mae)
 
 
+    def qrvs(self, size=None, d=None, qmc_engine=None):
```
Comment (mdhaber) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:1637]: This is indeed copy-pasted verbatim (aside from removing undesirable end of line whitespace)

File: scipy/stats/tests/test_sampling.py [hunk 4]
```diff
@@ -312,6 +312,100 @@ def check_discr_samples(rng, pv, mv_ex):
...
+        qrvs = gen.qrvs(size=size_in, d=d_in, qmc_engine=qrng)
+        if size_in is not None:
+            assert(qrvs.shape == shape_expected)
+
+        if qrng2 is not None:
+            uniform = qrng2.random(np.prod(size_in) or 1)
+            qrvs2 = stats.norm.ppf(uniform).reshape(shape_expected)
+            assert_allclose(qrvs, qrvs2, atol=1e-12)
+
+    def test_QRVS_size_tuple(self, method):
```
Comment (mdhaber) [inline scipy/stats/tests/test_sampling.py:375]: This was copy-pasted, generalized for `method` input.

File: scipy/stats/tests/test_sampling.py [hunk 5]
```diff
@@ -312,6 +312,100 @@ def check_discr_samples(rng, pv, mv_ex):
...
+    # `size=None` should not add anything to the shape, `size=1` should
+    sizes = [(None, tuple()), (1, (1,)), (4, (4,)),
+             ((4,), (4,)), ((2, 4), (2, 4))]  # type: ignore
+    # Neither `d=None` nor `d=1` should add anything to the shape
+    ds = [(None, tuple()), (1, tuple()), (3, (3,))]
+
+    @pytest.mark.parametrize('qrng', qrngs)
+    @pytest.mark.parametrize('size_in, size_out', sizes)
+    @pytest.mark.parametrize('d_in, d_out', ds)
+    def test_QRVS(self, qrng, size_in, size_out, d_in, d_out, method):
```
Comment (mdhaber) [inline scipy/stats/tests/test_sampling.py:347]: This was copy-pasted, generalized for method input.

File: scipy/stats/tests/test_sampling.py [hunk 6]
```diff
@@ -312,6 +312,100 @@ def check_discr_samples(rng, pv, mv_ex):
     assert pval > 0.1
 
 
+@pytest.mark.parametrize('method', ["NumericalInverseHermite", "NumericalInversePolynomial"])
+class TestQRVS:
+    def test_qrvs(self, method):
```
Comment (mdhaber) [inline scipy/stats/tests/test_sampling.py:317]: This looks the same, although a more descriptive/unique name (distinct from `test_QRVS`) below would be good. 

It looks like you cut out only parts related to QMC and left the rest behind, which seems right.

**Review by mdhaber — COMMENTED** [review]
I think @tirthasheshpatel should review and merge, but I don't find anything wrong with it other than the name of `test_qrvs`.

Since the `qrvs` methods are absolutely identical, I wonder whether they could inherit from a superclass.

Feel free to "Resolve Conversation" on the comments below. Just thought I'd document what I saw.

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ [hunk 7]
```diff
@@ -373,6 +373,26 @@ cdef double[::1] _validate_pv(pv) except *:
...
 
 
+def _validate_qmc_input(qmc_engine, d):
+    # Input validation for `qmc_engine` and `d`
+    # Error messages for invalid `d` are raised by QMCEngine
+    # we could probably use a stats.qmc.check_qrandom_state
+    if isinstance(qmc_engine, stats.qmc.QMCEngine):
+        message = "`d` must be consistent with dimension of `qmc_engine`."
+        if d is not None and qmc_engine.d != d:
+            raise ValueError(message)
```
Comment (tupui) [inline scipy/stats/_unuran/unuran_wrapper.pyx.templ:383]: Nitpick
```suggestion
        if d is not None and qmc_engine.d != d:
            message = "`d` must be consistent with dimension of `qmc_engine`."
            raise ValueError(message)
```

**Review by tupui — APPROVED** [review]
Thanks @chrisb83! Very happy to see another `qrvs`!

Mergeable as it is. I also think that we could move `qrvs` into `Method` as the code is similar. Or you wanted to add one by one and at the end do the change?

File: scipy/stats/tests/test_sampling.py (hunk 6 above)
Comment (chrisb83) [reply scipy/stats/tests/test_sampling.py]: I kept the name of the class and tried to improve the names of the methods

File: scipy/stats/_unuran/unuran_wrapper.pyx.templ (hunk 7 above)
Comment (chrisb83) [reply scipy/stats/_unuran/unuran_wrapper.pyx.templ]: fixed

**Review by tupui — APPROVED** [review]
Thanks for the update. LGTM then. @mdhaber feel free to merge if the update is ok with you.

**Comment by mdhaber** [comment]
It's fine but we have a test failure:
```python3
================================== FAILURES ===================================
________________________ stats/tests/test_sampling.py _________________________
[gw1] win32 -- Python 3.8.10 C:\hostedtoolcache\windows\Python\3.8.10\x86\python.exe
worker 'gw1' crashed while running 'stats/tests/test_sampling.py::TestQRVS::test_QRVS_shape_consistency[3-d_out2-4-size_out2-qrng1-NumericalInversePolynomial]'
```

**Comment by tupui** [comment]
> It's fine but we have a test failure:
> 
> ```python
> ================================== FAILURES ===================================
> ________________________ stats/tests/test_sampling.py _________________________
> [gw1] win32 -- Python 3.8.10 C:\hostedtoolcache\windows\Python\3.8.10\x86\python.exe
> worker 'gw1' crashed while running 'stats/tests/test_sampling.py::TestQRVS::test_QRVS_shape_consistency[3-d_out2-4-size_out2-qrng1-NumericalInversePolynomial]'
> ```

Indeed, with the recurrent timeout we are having I missed it. The message is weird

```
[gw1] node down: Not properly terminated
F
replacing crashed worker gw1
```

**Comment by chrisb83** [comment]
I ran the test again and it crashed with the same message as before...

File: scipy/stats/tests/test_sampling.py [hunk 8]
```diff
@@ -312,6 +312,100 @@ def check_discr_samples(rng, pv, mv_ex):
...
+                                    "NumericalInversePolynomial"])
+class TestQRVS:
+    def test_input_validation(self, method):
+        match = "`qmc_engine` must be an instance of..."
+        with pytest.raises(ValueError, match=match):
+            Method = getattr(stats.sampling, method)
+            gen = Method(StandardNormal())
+            gen.qrvs(qmc_engine=0)
+
+        if NumpyVersion(np.__version__) >= '1.18.0':
```
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_sampling.py:325]: We can remove this. The oldest support numpy is `1.18.5` in SciPy 1.9.0.

File: scipy/stats/tests/test_sampling.py [hunk 9]
```diff
@@ -312,6 +312,100 @@ def check_discr_samples(rng, pv, mv_ex):
...
+            # issues with QMCEngines and old NumPy
+            Method = getattr(stats.sampling, method)
+            gen = Method(StandardNormal())
+
+            match = "`d` must be consistent with dimension of `qmc_engine`."
+            with pytest.raises(ValueError, match=match):
+                gen.qrvs(d=3, qmc_engine=stats.qmc.Halton(2))
+
+    if NumpyVersion(np.__version__) >= '1.18.0':
```
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_sampling.py:334]: Same here.

File: scipy/stats/tests/test_sampling.py [hunk 10]
```diff
@@ -312,6 +312,100 @@ def check_discr_samples(rng, pv, mv_ex):
...
+             ((4,), (4,)), ((2, 4), (2, 4))]  # type: ignore
+    # Neither `d=None` nor `d=1` should add anything to the shape
+    ds = [(None, tuple()), (1, tuple()), (3, (3,))]
+
+    @pytest.mark.parametrize('qrng', qrngs)
+    @pytest.mark.parametrize('size_in, size_out', sizes)
+    @pytest.mark.parametrize('d_in, d_out', ds)
+    def test_QRVS_shape_consistency(self, qrng, size_in, size_out,
+                                    d_in, d_out, method):
+        dist = StandardNormal()
```
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_sampling.py:349]:
```suggestion
        if platform.architecture()[0] == "32bit":
            pytest.xfail("NumericalInversePolynomial.qrvs fails for "
                         "32-bit windows")
        dist = StandardNormal()
```

**Review by tirthasheshpatel — APPROVED** [review]
Hi @chrisb83! Sorry for the very late review! I am trying to reproduce the error locally but I am not able to get a 32-bit build working on windows. I also tried in CI but that keeps failing too. How about we xfail on 32-bit windows for now and revisit this later?

Other than that, I left a few non-blocking comments. The `qrvs` method looks good and tests look very through. Thanks for working on this @chrisb83!

File: scipy/stats/tests/test_sampling.py (hunk 8 above)
Comment (tupui) [reply scipy/stats/tests/test_sampling.py:325]: Good catch, I missed it as it was a move!

**Comment by tirthasheshpatel** [comment]
Azure doesn't allow me to re-run failed jobs:

image

(btw i signed in using github)

**Comment by chrisb83** [comment]
> Azure doesn't allow me to re-run failed jobs:

I restarted the job. Reason for the fail was u related: the mingw installation failed

**Comment by tupui** [comment]
> Azure doesn't allow me to re-run failed jobs:
> 
> image
> 
> (btw i signed in using github)

Yes only some users have access (I don't and I am not sure who has access), it's not linked to GH and any orgs here like CircleCi.

**Comment by tupui** [comment]
ASV failure is unrelated (it's the version issue, so merging main would fix it).

**Comment by chrisb83** [comment]
Bad news: " worker 'gw0' crashed while running 'stats/tests/test_sampling.py::TestQRVS::test_QRVS_shape_consistency[3-d_out2-4-size_out2-None-NumericalInversePolynomial]'"

**Comment by tirthasheshpatel** [comment]
I pushed a [new commit](https://github.com/tirthasheshpatel/scipy/runs/5381485315?check_suite_focus=true) on my fork to test with Python 3.8. Let's see if that fails. If not, I would say xfail the test in CI and revisit it later (I suspect it's a bug in UNU.RAN since Python debug build doesn't complain).

**Comment by tupui** [comment]
Fine with me if we cannot reproduce it. (FYI, you can force a test to xfail, this way the job would fail if the test passes. A way to know something changed and keep track of this. OC only works if the fail is consistent.)

**Comment by mdhaber** [comment]
`xfail` sounds good. @tupui you mean `@pytest.mark.xfail(strict=True)`, right? That sounds good if the failure is consistent, yes.

**Comment by tupui** [comment]
Yes exactly!

**Comment by chrisb83** [comment]
Brief status update:

- marking the test as xfail with `strict=True` does not work since the test is parametrized and even on Windows 32, only one of the many test cases fails. the remaining ones would trigger a fail if they pass and strict=True. so I follow the approach suggested by Tirth to xfail the test during execution
- to identify the Windows 32bit build, I use `sys.platform == "win32" and platform.architecture()[0] == "32bit"`. `sys.platform` might apparenly return win32 even if it is a 64bit system (https://stackoverflow.com/questions/23956229/i-have-a-win7-64bits-os-but-the-value-of-sys-platform-is-win32-why).

Let me know if you prefer a different approach. Otherwise I will fix the remaining lint error within the next days

File: scipy/stats/tests/test_sampling.py [hunk 11]
```diff
@@ -312,6 +314,98 @@ def check_discr_samples(rng, pv, mv_ex):
...
+    ds = [(None, tuple()), (1, tuple()), (3, (3,))]
+
+    @pytest.mark.parametrize('qrng', qrngs)
+    @pytest.mark.parametrize('size_in, size_out', sizes)
+    @pytest.mark.parametrize('d_in, d_out', ds)
+    def test_QRVS_shape_consistency(self, qrng, size_in, size_out,
+                                    d_in, d_out, method):
+        w32 = sys.platform == "win32" and platform.architecture()[0] == "32bit"
+        if w32 and method == "NumericalInversePolynomial":
+            pytest.xfail("NumericalInversePolynomial.qrvs fails for Win 32-bit")
```
Comment (tirthasheshpatel) [inline scipy/stats/tests/test_sampling.py:349]:
```suggestion
            pytest.xfail("NumericalInversePolynomial.qrvs fails for Win "
                         "32-bit")
```

To resolve the lint failure

**Review by tirthasheshpatel — APPROVED** [review]
All tests passing now! Once UNU.RAN 1.9.0 is released (which should be soon) and we update scipy/unuran, it would be good to check if the 32-bit windows failure is resolved. But, for now, this looks good to me. Tests are comprehensive and strong (thanks @mdhaber). The implementation looks correct.

I think it's safe to merge as this has been up for a few weeks without any blocking comments (modulo the test failure). Thanks @chrisb83, @tupui, @mdhaber for all the reviews and work on this!!

PS: There's a bit of duplication in `unuran_wrapper.pyx.templ` now and the file has become huge. I will clean it up in a follow-up.

**Comment by chrisb83** [comment]
Thanks for your help, Tirth, Matt and Pamphile!

**Comment by tupui** [comment]
Very happy about this feature! Thanks all.

**Pull request merged**